### PR TITLE
[WIP] Use user supplied S3 instance

### LINKED ISSUE #9: Drop requirements for aws-sdk data and dirname
As `aws-sdk` auto-loads configurations and credentials from `~/.aws/` directory and the [recommended](http://docs.aws.amazon.com/AWSJavaScriptSDK/guide/node-configuring.html) way of loading credentials is by letting the `aws-sdk` do so, I suggest we drop the requirements for the region, secretAccessKey and accessKeyId.

The dirname option also should not be required, as it is not mandatory to upload the file to a directory (though useful for most cases).

What do you think?
```
  if (!opts.secretAccessKey) throw new Error('secretAccessKey is required')
  if (!opts.accessKeyId) throw new Error('accessKeyId is required')
  if (!opts.region) throw new Error('region is required')
  if (!opts.dirname) throw new Error('dirname is required')
```

### PR DESCRIPTION
**Pull request by LinusU**
This changes the functionality to use a user supplied `s3` instance (`aws.S3`) instead of creating our own. That way you are free to authenticate in any way you like, plus set any additional options you may want.

Fixes #9 

The tests probably could use some love, the good thing is that we no longer need to mock a server, but instead just the s3 instance. Take a look at the test code to see this in action.

I want to test all available options of all options and also try it with some different files. Another thing I noticed is that I reuse much of the code from `multer`s test set, we should probably move all that to a separate package on npm and require from both parts.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,9 +1,13 @@
-var aws = require('aws-sdk')
-var extend = require('xtend')
 var crypto = require('crypto')
 var stream = require('stream')
 var fileType = require('file-type')
 
+function staticVaule (value) {
```
Comment (badunk) [inline index.js]: typo, should be staticValue

File: index.js [hunk 2]
```diff
@@ -27,53 +31,86 @@ function autoContentType (req, file, cb) {
...
-  if (!opts.region) throw new Error('region is required')
+function collect (storage, req, file, cb) {
+  storage.getBucket(req, file, function (err, bucket) {
+    if (err) return cb(err)
+
+    storage.getKey(req, file, function (err, key) {
+      if (err) return cb(err)
 
-  var s3cfg = extend(opts, { apiVersion: '2006-03-01' })
+      storage.getContentType(req, file, function (err, contentType, replacementStream) {
```
Comment (badunk) [inline index.js]: why is the replacementStream passed through `contentType` signature?

File: index.js [hunk 3]
```diff
@@ -27,53 +31,86 @@ function autoContentType (req, file, cb) {
...
+    })
+  })
+}
 
-  this.options = opts
-  this.getKey = (opts.key || defaultKey)
-  this.getContentType = (opts.contentType || defaultContentType)
-  this.s3 = new aws.S3(s3cfg)
+function S3Storage (opts) {
+  switch (typeof opts.s3) {
```
Comment (badunk) [inline index.js]: awesome way to validate

**Comment by badunk** [comment]
I made a PR at #26 with some small fixes

**Comment by LinusU** [comment]
Thanks man!! Sorry I dropped the ball on this one

**Comment by badunk** [comment]
Hey no problem at all - you see like a busy guy ;)

**Pull request closed without merging**